backend/app.js

```diff
@@ -78,46 +45,62 @@ wss.on("connection", (ws) => {
       return;
     }
 
-    if (clients[url].find((w) => w === ws) === undefined) {
+    if (clients[url].find((c) => c.ws === ws) === undefined) {
       console.error("Drawing action to incorrect url");
       return;
     }
 
     if (event.action === "DRAWING_EVENTS") {
-      addEvent(url, JSON.stringify(event));
+      addDrawEvent(url, JSON.stringify(event));
 
-      clients[url].forEach((client) => {
+      clients[url].forEach((obj) => {
+        const client = obj.ws;
         if (client !== ws && client.readyState === WebSocket.OPEN) {
           client.send(JSON.stringify(event));
         }
       });
     } else if (event.action === "REQUEST_DRAWING_HISTORY") {
-      getAllEvents(url, (drawingHistoryString) => {
+      getAllDrawEvents(url, (drawingHistoryString) => {
         const drawingHistory = JSON.parse(drawingHistoryString);
         const payload = JSON.stringify({
           action: "DRAWING_HISTORY",
           drawingHistory
         });
         ws.send(payload);
       });
+    } else if (event.action === "WRITE_EVENT") {
+      addWriteEvent(url, JSON.stringify(event.event));
     }
   });
 });
 
+const intervalTimeout = setInterval(() => {
```

Comment:
If I understand this correctly, the backend periodically sends writing history to clients while drawing history is immediately broadcasted after being received by the server. Any particular reason why this is better off behaving differently?

Reply:
Hmm, could this immediately instead.

The main difference is that I keep the DB as the source of the correct ordering of events and resend events received from a client to that client to ensure (eventual) consistency between clients. The idea is that the write actions are idempotent --- as long as the actions are done in the correct order eventually the state will be correct. When received, the actions are added to the DB and that order is then the "truth".

So I could do this in the receive function. However, I would still first add to the DB, then fetch from DB, then send

Reply:
Actually, I think I'll do that